bundles/org.openhab.binding.heliosventilation/src/main/java/org/openhab/binding/heliosventilation/internal/HeliosVentilationHandler.java

```diff
@@ -0,0 +1,454 @@
+
+    }
+
+    private void poll(HeliosVentilationDataPoint v) {
+        byte txFrame[] = { 0x01, BUSMEMBER_ME, BUSMEMBER_MAINBOARD, 0x00, v.address(), 0x00 };
+        txFrame[5] = (byte) checksum(txFrame);
+
+        tx(txFrame);
+    }
+
+    /*
+     * transmit a frame
+     */
+    private void tx(byte[] txFrame) {
+        try {
+            OutputStream out = outputStream;
+            if (out != null) {
+                if (logger.isTraceEnabled()) {
+                    logger.trace("HeliosVentilation: Write to serial port: {}",
+                            String.format("%02x %02x %02x %02x", txFrame[1], txFrame[2], txFrame[3], txFrame[4]));
+                }
+
+                out.write(txFrame);
+                out.flush();
+                // after each frame we have to wait.
+                // 30 ms is taken from what we roughly see the original remote control is doing
+                Thread.sleep(30);
+            }
+        } catch (IOException e) {
+            // in case we cannot write the connection is somehow broken, let's officially disconnect
+            disconnect();
+            connect();
+        } catch (InterruptedException e) {
+            // ignore if we got interrupted
+        }
+    }
+
+    /**
+     * Check connection status
+     *
+     * @return true if currently connected
+     */
+    private boolean isConnected() {
+        return serialPort != null && inputStream != null && outputStream != null;
+    }
+
+    @Override
+    public synchronized void serialEvent(SerialPortEvent event) {
+        switch (event.getEventType()) {
+            case SerialPortEvent.DATA_AVAILABLE:
+                // we get here if data has been received
+
+                try {
+                    // Wait roughly a frame length to increase the likelihood that the complete frame is already
+                    // buffered. This improves the robustness for RS485/USB converters which sometimes duplicate bytes
+                    // otherwise
+                    Thread.sleep(8);
+                } catch (InterruptedException e) {
+                    // ignore interruption
+                }
+
+                byte frame[] = { 0, 0, 0, 0, 0, 0 };
+                InputStream in = inputStream;
+                if (in != null) {
+                    try {
+                        do {
+                            int cnt = 0;
+                            do {
+                                // read data from serial device
+                                while (cnt < 6 && in.available() > 0) {
+                                    final int bytes = in.read(frame, cnt, 1);
+                                    if (cnt > 0 || frame[0] == 0x01) {
+                                        // only proceed if the first byte was 0x01
+                                        cnt += bytes;
+                                    }
+                                }
+                            } while (in.available() > 0 && cnt < 6);
```

Comment:
it doesn't look like this outer while loop serves any purpose.

Reply:
in case there are already more than 6 bytes read, the outer loop is there to start reading directly the next (already partially received) frame

Reply:
ah no, you mean the "inner" outer loop :-) that one is useless I agree.